### PR TITLE
ELE-6471 Round timestamps if they exist

### PR DESCRIPTION
**Pull request by maddocash**
- talis/elevate-app#6471

There is a proposed change to the echo api to only accept integer timestamp values. We will try to ensure we only send interger values but to catch any events we might miss we're adding a step in the echo client (used by all the calls in elevate to echo) to round the timestamp to the nearest whole number.

### CODE REVIEW

File: echo/test/unit/echo_client_test.js
Comment (mike-dean-talis) [inline echo/test/unit/echo_client_test.js]:
```suggestion
                (err === null).should.be.true();
```

File: echo/test/unit/echo_client_test.js
Comment (mike-dean-talis) [inline echo/test/unit/echo_client_test.js]:
```suggestion
                (err === null).should.be.true();
```

File: echo/test/unit/echo_client_test.js
Comment (mike-dean-talis) [inline echo/test/unit/echo_client_test.js]: suggestion: I wonder if you should include another event with a timestamp of `1324524510.673` and check the result is rounded to the nearest integer (rather than being `floored` for example)?

File: echo/test/unit/echo_client_test.js
Comment (maddocash) [reply echo/test/unit/echo_client_test.js]: I copied this from the other tests and I think it's ok?
https://github.com/tj/should.js/#true

Comment (mike-dean-talis) [reply echo/test/unit/echo_client_test.js]: This looks like old documentation to me?

http://shouldjs.github.io/#assertion-true

As far as I understand it, `should` assertions need to be methods. Hence installing a linter in Elevate to ensure this happens.

Comment (maddocash) [reply echo/test/unit/echo_client_test.js]: I agree there is an inconsistency with their readme and their docs site. I have no preference either way so I was follow the 'house style'. So the question is do we want to add in consistency here? What is the reason for preferring the method call?

Comment (mike-dean-talis) [reply echo/test/unit/echo_client_test.js]: Change this line to `(err === null).should.be.false`.

It will pass.

This is incorrect.

Comment (mike-dean-talis) [reply echo/test/unit/echo_client_test.js]: In order to ensure the statement is asserted, you must call the method (so `true()` or `false()`), otherwise this is effectively a noop.

**Review by steveellisdev — APPROVED** [review]

**Review by mike-dean-talis — APPROVED** [review]

File: echo/test/unit/echo_client_test.js
Comment (ShaunParsons) [inline echo/test/unit/echo_client_test.js:243]: Whats the behaviour if the `timestamp` is not set?

**Review by ShaunParsons — CHANGES_REQUESTED** [review]
I don't think the client library should be modifying the event data. That might be unexpected/undesired for other consumers of this library. Instead I think we should be throwing an error if the timestamp is invalid, which is exactly what we'll be doing with the API shortly.

File: echo/test/unit/echo_client_test.js
Comment (maddocash) [reply echo/test/unit/echo_client_test.js:243]: it does nothing, I'll add a test, if we agree on the changes here

**Comment by steveellisdev** [comment]
> I don't think the client library should be modifying the event data. That might be unexpected/undesired for other consumers of this library. Instead I think we should be throwing an error if the timestamp is invalid, which is exactly what we'll be doing with the API shortly.

The problem with just throwing an error is that means we then lose data anywhere we haven't caught this conversion change, doing it in the library is a much safer way of doing it. I agree that in an ideal world the client would just be passing the data across in the correct format, but as you've found we're sending event data from all over the place, in different apps, and it will be very easy to miss somewhere and then lose data. If we're expecting the ids passed in to be integers then I don't see why rounding that would cause unexpected results.

Doing this in Elevate is a lot more work as we'll pretty much have to audit everywhere we're sending echo events off, we won't have the capacity to do that until at least Q1 as our focus is now only the new player. That work can still go ahead even with this change as I agree that Elevate shouldn't be trying to send off invalid timestamps. So the question is whether we care enough about this to hold up the wider project until next year around adding Echo validation until we can do that audit.

**Comment by malcyL** [comment]
If we want to do this in one place, what about:

- echo continues to accept the timestamp, for now
- echo secures the exact copy of the event in the global store - so there is never a question of what the original data is, should something, and I quote "unexpected/undesired" happen.
- the transformation / data clean up being done in this PR is done as the data is loaded into the time based store This would be trivial to do, Kinesis, Firehose etc all have the ability to transform the data. 

Advantages I see:

- No time pressure on elevate or aspire to do this work
- Having secured a copy of exactly what is sent, anything "unexpected/undesired" is at least traceable and fixable
- This is done in one place, catching all of elevates events without the possibility of, as Steve says, loosing data.

**Comment by steveellisdev** [comment]
That approach sounds good to me and sounds like the best way to completely unblock things from your side. I'll create a ticket to do the audit on our side and add it to our tech debt project.

**Pull request closed without merging**
